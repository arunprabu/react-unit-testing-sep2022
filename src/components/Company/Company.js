import React, { useState } from 'react'

const Company = ( {name} ) => {

  console.log('Program Started');

  const [country, setCountry] = useState('Canada');

  const handleCountryChange = (event) => {
    console.log(event); // event obj 
    console.log(event.target.value);
    setCountry(event.target.value);
  }

  return (
    <div>
      <h2>Company | Testing Props, Event, States, Hooks, JSX, CSS styles and Snapshot</h2>
      <h3 data-testid="companyNameEl">
        Cognizant's Official Website
      </h3>

      <h4>
        Company Name: 
        <p data-testid="companyFullName">{name}</p>
      </h4>

      <br />
      <label>Enter your Country: </label>
      {/* working with onChange event */}
      <input type='text' value={country} 
        onChange={handleCountryChange}
        placeholder="Enter Country"/>
      <p>Head over to Cognizant {country} website!</p>

    </div>
  )
}

export default Company;