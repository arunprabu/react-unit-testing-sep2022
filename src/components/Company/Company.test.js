// Arrange
import { fireEvent, render, screen } from "@testing-library/react";
import renderer from 'react-test-renderer';
import Company from "./Company";

// TEST SUITE
describe('CompanyComponent', () => {

  // test spec #1
  it('has the company name Cognizant', () => {
    render(<Company />);
    const companyNamEl = screen.getByTestId('companyNameEl');

    expect(companyNamEl.textContent)
      .toBe(`Cognizant's Official Website!`);
  });

  // test spec #2 -- let's test the props
  it('receive props and displays them properly', () => {
    render(<Company name='Cognizant Technology Solutions Corporation'/>);
    
    expect(screen.getByTestId('companyFullName')).toHaveTextContent('Cognizant Technology Solutions Corporation');
  });

  // test spec #3
  // checking if an input element with 'Enter Country' as placeholder found or not 
  it(`has an input element with 'Enter Country' as placeholder`, () => {
    render(<Company />);
    const countryInput = screen.getByPlaceholderText('Enter Country');
    expect(countryInput).toBeTruthy();
  });

  // test spec #4 
  it('should update country onChange event', ()=> {
    render(<Company />);
    // let's find the input element that has onChange handler
    const countryInput = screen.getByPlaceholderText('Enter Country');
    
    // now checking if onChange event is working well or not
    // mock fire the onChange event with mock value
  
    fireEvent.change(countryInput, {
      target: {
        value: 'Australia'
      }
    });
    expect(countryInput.value).toBe('Australia');
    expect(screen.getByText('Head over to Cognizant Australia website!')).toBeInTheDocument();

    fireEvent.change(countryInput, {
      target: {
        value: 'UK'
      }
    });
    expect(countryInput.value).toBe('UK');
    expect(screen.getByText('Head over to Cognizant UK website!')).toBeInTheDocument();
  });

  // testing inline css 
  it('has css style with bg color green', () => {
    render(<Company />);
    expect(screen.getByTestId('successNotification')).toHaveStyle('background-color: #00ff00' );
  });

  // TODO: testing element with css class 'btn-danger' 

  // snapshot testing
  it('should have right company comp snapshot', () => {
    // npm i react-test-renderer 
    // we will take a snapshot in json format
    const snapshotJSON = renderer.create(<Company name='Cognizant Technology Solutions Corporation'/> );
    // lets assert
    expect(snapshotJSON).toMatchSnapshot();
  });



});

