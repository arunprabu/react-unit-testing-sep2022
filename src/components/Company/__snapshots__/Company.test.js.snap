// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompanyComponent should have right company comp snapshot 1`] = `
<div>
  <h2>
    Company | Testing Props, Event, States, Hooks, JSX, CSS styles and Snapshot
  </h2>
  <h3
    data-testid="companyNameEl"
  >
    Cognizant's Official Website!
  </h3>
  <h4>
    Company Name:
    <p
      data-testid="companyFullName"
    >
      Cognizant Technology Solutions Corporation
    </p>
  </h4>
  <br />
  <label>
    Enter your Country: 
  </label>
  <input
    onChange={[Function]}
    placeholder="Enter Country"
    type="text"
    value="Canada"
  />
  <p>
    Head over to Cognizant 
    Canada
     website!
  </p>
  <br />
  <div
    data-testid="successNotification"
    style={
      Object {
        "backgroundColor": "#00ff00",
        "color": "#000",
        "padding": "20px",
      }
    }
  >
    Success! Data Saved!!
  </div>
</div>
`;
